VariantShredding.md

```diff
@@ -25,276 +25,302 @@
+  required binary metadata;
+  optional binary value;                # a variant, expected to be an object
+  optional group typed_value {          # shredded fields for the variant object
+    required group event_type {         # shredded field for event_type
+      optional binary value;
+      optional binary typed_value (STRING);
+    }
+    required group event_ts {           # shredded field for event_ts
+      optional binary value;
+      optional int64 typed_value (TIMESTAMP(true, MICROS));
+    }
+  }
+}
 ```
 
-# Nested Parquet Example
+The group for each named field must be required.
 
-This section describes a more deeply nested example, using a top-level array as the shredding type.
+A field's `value` and `typed_value` are set to null (missing) to indicate that the field does not exist in the variant.
+To encode a field that is present with a null value, the `value` must contain a Variant null: basic type 0 (primitive) and physical type 0 (null).
 
-Below is a sample of JSON that would be fully shredded in this example.
-It contains an array of objects, containing an `a` field shredded as an array, and a `b` field shredded as an integer.
+The series of objects below would be stored as:
 
-```
-[
-  {
-    "a": [1, 2, 3],
-    "b": 100
-  },
-  {
-    "a": [4, 5, 6],
-    "b": 200
-  }
-]
-```
+| Event object                                                                       | `value`                           | `typed_value` | `typed_value.event_type.value` | `typed_value.event_type.typed_value` | `typed_value.event_ts.value` | `typed_value.event_ts.typed_value` | Notes                                         |
+|------------------------------------------------------------------------------------|-----------------------------------|---------------|--------------------------------|--------------------------------------|------------------------------|------------------------------------|-----------------------------------------------|
+| `{"event_type": "noop", "event_ts": 1729794114937}`                                | null                              | non-null      | null                           | `noop`                               | null                         | 1729794114937                      | Fully shredded object                         |
+| `{"event_type": "login", "event_ts": 1729794146402, "email": "user@example.com"}`  | `{"email": "user@example.com"}`   | non-null      | null                           | `login`                              | null                         | 1729794146402                      | Partially shredded object                     |
+| `{"error_msg": "malformed: ..."}`                                                  | `{"error_msg", "malformed: ..."}` | null          |                                |                                      |                              |                                    | Object with no shredding                      |
```

Comment:
I think `typed_value` should be non-null (but still with all-null children). If a query needs to fetch `struct<event_type, event_ts>`, it can be done without needing the `value` column, and will still produce the correct struct nullness.

Based on @RussellSpitzer's question above, it also might be good to add an example of an empty object (`{ }`), where again, I think `typed_value` would be non-null, but `value` would be null.

Reply:
I think this is kind of continuation of the above example. Having typed_value null looks good to me.

Reply:
Updated.